### PR TITLE
feature: download renderer to facilitate file downloads

### PR DESCRIPTION
**Pull request by pkgmain**
This is what I'm using for file downloads in my projects so I thought I'd open a PR to see if there's any desire for it in Buffalo proper. Example usage:

```go
c.Render(http.StatusOK, r.Download(c, "report.pdf", bytes.NewReader(pdfBytes))
```

### CODE REVIEW

File: render/download.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	t := mime.TypeByExtension(ext)
+	if t == "" {
+		t = "application/octet-stream"
+	}
+
+	return t
+}
+
+func (r downloadRenderer) Render(w io.Writer, d Data) error {
+	r.writer.Header().Add("Content-Disposition", fmt.Sprintf("attachment; filename=%s", r.filename))
```
Comment (markbates) [inline render/download.go]: Cast w if it’s an http writer,or better yet it’s a writer that implements an interface that has Header() then write the headers.

File: render/download.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+// Download renders a file attachment automatically setting following headers:
+//
+//   Content-Type
+//   Content-Length
+//   Content-Disposition
+//
+// Content-Type is set using mime#TypeByExtension with the filename's extension. Content-Type will default to
+// application/octet-stream if using a filename with an unknown extension.
+func Download(data []byte, filename string, writer http.ResponseWriter) Renderer {
```
Comment (markbates) [inline render/download.go]: Args should be name string and r Reader

File: render/download.go (hunk 1 above)
Comment (markbates) [reply render/download.go]: Yes, sorry, you're right. That's what happens when you do PR reviews before coffee. :)

File: render/download.go (hunk 2 above)
Comment (markbates) [reply render/download.go]: In that case you should change the signature model to look more like: https://github.com/gobuffalo/buffalo/blob/master/render/auto.go#L43

```go
func Download(ctx buffalo.Context, name string, r io.Reader) Renderer
```

Comment (markbates) [reply render/download.go]: then you have everything from the context, so you can do whatever you want and you can use anything for `io.Reader`.

Comment (pkgmain) [reply render/download.go]: I think using a `buffalo.Context` would cause a cyclical import, no?

Comment (markbates) [reply render/download.go]: d'oh, I meant `context.Context`, like here:
https://github.com/gobuffalo/buffalo/blob/master/render/auto.go#L43

Then you can check against a private interface:

```go
type responsible interface {
  Response() http.ResponseWriter
}

func Download(ctx context.Context, name string, r io.Reader) Renderer {
  bc, ok := ctx.(responsible)
  if ok {
// ...
  }
// ...
}
```

Something like that anyway. That way no cyclical imports, easily testable, etc... hopefully.

Comment (pkgmain) [reply render/download.go]: Gotcha, that makes sense. Updated.

**Pull request merged**